mypy/checker.py

```diff
@@ -4587,6 +4763,75 @@ def or_conditional_maps(m1: TypeMap, m2: TypeMap) -> TypeMap:
     return result
 
 
+def or_partial_conditional_maps(m1: TypeMap, m2: TypeMap) -> TypeMap:
```

Comment:
This name doesn't really reflect what this does. Maybe use `combine_conditional_maps()` would be better? I however don't have strong feelings here.

Reply:
I think using the word "combine" would be a bit ambiguous because it'd be unclear whether the combining step will end up "and"-ing or "or"-ing the the two maps.

Basically, the way I was thinking about this change is that there are now two kinds of TypeMaps -- full ones, which represent all info we know about types within a certain "context", and partial ones which contain only some of the info.

Partial TypeMaps also only exist as an implementation detail of the narrowing logic: it always ends up returning full TypeMaps.

So if `and_conditional_maps` and `or_conditional_maps` functions are for full TypeMaps, I was thinking it made sense to add `and_partial_conditional_maps` and `or_partial_conditional_maps` functions for the partial ones.

But it *also* turned out that `and_conditional_maps` and `and_partial_conditional_maps` behave in the exact same way/would share the same implementation, so I didn't bother defining the latter.

The other approach I could take is to *not* define the `or_partial_conditional_maps` and instead just rewrite `reduce_partial_conditional_maps` function to produce the output maps more directly and do the (pseudo)-intersecting and unioning itself.

That works just as well (and is actually probably slightly more efficient), but does end up making it harder to see how the function relates to the existing conditional map logic.